practice_2/WeatherApp/VIew/WeatherViewController.swift

```diff
@@ -0,0 +1,146 @@
+        initRecognizer()
+        initSearchController()
+    }
+
+    private func setup() {
+        let networkService = WeatherNetworkService.sharedWeather
+        self.presenter = WeatherPresenter(view: self, networkService: networkService)
+    }
+
+    private func initRecognizer() {
+        let gestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(tapAnnotationAction))
+        mapView.addGestureRecognizer(gestureRecognizer)
+    }
+
+    private func initSearchController() {
+        let storyboard = UIStoryboard(name: "WeatherApp", bundle: nil)
+        guard let locationSearchTable = storyboard.instantiateViewController(
+                identifier: "LocationSearchTable") as? LocationSearchTable else { return }
+        resultSearchController = UISearchController(searchResultsController: locationSearchTable)
+        resultSearchController?.searchResultsUpdater = locationSearchTable
+        locationSearchTable.mapView = mapView
+        locationSearchTable.handleMapSearchDelegate = self
+
+        let searchBar = resultSearchController?.searchBar
+        searchBar?.sizeToFit()
+        searchBar?.placeholder = "Search for places"
+        navigationItem.titleView = resultSearchController?.searchBar
+
+        resultSearchController?.hidesNavigationBarDuringPresentation = false
+        resultSearchController?.obscuresBackgroundDuringPresentation = true
+        definesPresentationContext = true
+    }
+
+    @objc func tapAnnotationAction(gestureRecognizer: UIGestureRecognizer) {
+        let location = getLocation(from: gestureRecognizer)
+        let tempUnits = getTemperatureUnits()
+        presenter?.fetchWeather(by: location, units: tempUnits.urlUnits)
+        presenter?.fetchForecast(by: location, units: tempUnits.urlUnits)
+        let temperature = presenter?.weather?.temperature
+        addAnnotation(by: location, temperature: temperature, unit: tempUnits.unit)
+    }
+
+    private func getTemperatureUnits() -> (unit: String, urlUnits: String) {
+        if let unit = UserDefaults.standard.string(forKey: "temperutureUnit") {
+            if unit == "°C" {
+                return (unit: "°C", urlUnits: "metric")
+            } else {
+                return (unit: "°F", urlUnits: "imperial")
+            }
+        } else {
+            return defaultUnits
+        }
+    }
+
+    private func getLocation(from gestureRecognizer: UIGestureRecognizer) -> CLLocationCoordinate2D {
+        let touchLocation = gestureRecognizer.location(in: mapView)
+        let location = mapView.convert(touchLocation, toCoordinateFrom: mapView)
+        return location
+    }
+
+    private func addAnnotation(by location: CLLocationCoordinate2D, temperature: Double?, unit: String) {
+        let annotation = MKPointAnnotation()
+        if let temperature = temperature {
+            annotation.title = "\(temperature) \(unit)"
+        } else {
+            annotation.title = "\(temperature as Double?)"
```

Comment:
add error handling in as Double conversion

Reply:
Handled that point in another way.
Please have a look. Thanks.